processors/metricresourceattrstoattrs/metricresourcesattrstolabels.go

```diff
@@ -40,38 +40,103 @@ func (p *processor) ProcessMetrics(ctx context.Context, metrics pmetric.Metrics)
-						for l := 0; l < metricData.Len(); l++ {
-							metricData.At(l).Attributes().Insert(key, v)
-						}
-					case pmetric.MetricDataTypeSum:
-						metricData := metric.Sum().DataPoints()
-						for l := 0; l < metricData.Len(); l++ {
-							metricData.At(l).Attributes().Insert(key, v)
-						}
-					case pmetric.MetricDataTypeHistogram:
-						metricData := metric.Histogram().DataPoints()
-						for l := 0; l < metricData.Len(); l++ {
-							metricData.At(l).Attributes().Insert(key, v)
-						}
-					case pmetric.MetricDataTypeExponentialHistogram:
-						metricData := metric.ExponentialHistogram().DataPoints()
-						for l := 0; l < metricData.Len(); l++ {
-							metricData.At(l).Attributes().Insert(key, v)
-						}
-					case pmetric.MetricDataTypeSummary:
-						metricData := metric.Summary().DataPoints()
-						for l := 0; l < metricData.Len(); l++ {
-							metricData.At(l).Attributes().Insert(key, v)
-						}
-					}
+					addResourceAttributeToMetricAttributes(metric, key, v)
 					return true
 				})
+				removeJobAndInstanceLabels(metric, hasResourceServiceNameAttr, hasResourceServiceInstanceIDAttr)
 			}
 		}
 	}
 	return metrics, nil
 }
+
+// addResourceAttributeToMetricAttributes is the workhorse to add resource attributes to metric attributes
+func addResourceAttributeToMetricAttributes(metric pmetric.Metric, key string, v pcommon.Value) {
+	metricDataType := metric.DataType()
+	switch metricDataType {
+	case pmetric.MetricDataTypeGauge:
+		metricData := metric.Gauge().DataPoints()
+		for l := 0; l < metricData.Len(); l++ {
+			metricData.At(l).Attributes().Insert(key, v)
+		}
+	case pmetric.MetricDataTypeSum:
+		metricData := metric.Sum().DataPoints()
+		for l := 0; l < metricData.Len(); l++ {
+			metricData.At(l).Attributes().Insert(key, v)
+		}
+	case pmetric.MetricDataTypeHistogram:
+		metricData := metric.Histogram().DataPoints()
+		for l := 0; l < metricData.Len(); l++ {
+			metricData.At(l).Attributes().Insert(key, v)
+		}
+	case pmetric.MetricDataTypeExponentialHistogram:
+		metricData := metric.ExponentialHistogram().DataPoints()
+		for l := 0; l < metricData.Len(); l++ {
+			metricData.At(l).Attributes().Insert(key, v)
+		}
+	case pmetric.MetricDataTypeSummary:
+		metricData := metric.Summary().DataPoints()
+		for l := 0; l < metricData.Len(); l++ {
+			metricData.At(l).Attributes().Insert(key, v)
+		}
+	}
+}
+
+// removeJobAndInstanceLabels removes model.JobLabel if hasResourceServiceNameAttr and model.InstanceLabel if hasResourceServiceInstanceIDAttr,
+// from the metric attributes. This is because they will be added by the prometheus exporter based on serviceName and serviceInstanceId
+// and if already present they will be duplicate and will cause an error while processing metrics.
+func removeJobAndInstanceLabels(metric pmetric.Metric, hasResourceServiceNameAttr bool, hasResourceServiceInstanceIDAttr bool) {
+	metricDataType := metric.DataType()
+	switch metricDataType {
+	case pmetric.MetricDataTypeGauge:
```

Comment:
this switch/pluck/iterate/access attributes is duplicate both across the add/remove functions and each type case. I don't know golang well enough to say exactly how, but you should be able to abstract it and shrink the body of the add/remove functions down to a few lines each - just the two if cases in the remove case, the insert in the add.

Reply:
will rework this.